concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -586,63 +719,64 @@ writeChunks
                         outDir
                         chunkSize
                         Nothing
+                        lastFinalizationRecordIndex'
+                        getBlockAt
+                        getFinalizationRecordAt
                 return $ chunkInfo :<| chunks
             else return $ singleton chunkInfo
 
 -- |Export a series of blocks as a chunk of a specified length. For each block containing a
 -- finalization record, the 'dbsLastFinIndex' field of the state is updated with its finalization
 -- index.
 exportBlocksToChunk ::
-    forall pv m.
-    (IsProtocolVersion pv, MonadIO m, MonadState (DBState pv) m, MonadCatch m) =>
+    (MonadIO m) =>
     -- |Handle to export to
     Handle ->
     -- |Height of next block to export
     BlockHeight ->
     -- |Number of blocks to export
     Word64 ->
-    -- |Number of exported blocks
-    m Word64
-exportBlocksToChunk hdl firstHeight chunkSize = ebtc firstHeight 0
+    -- |Last finalization record index.
+    FinalizationIndex ->
+    -- |Action for getting the block.
+    GetBlockAt m ->
+    -- |Number of exported blocks and last finalization record index.
+    m (Word64, FinalizationIndex)
+exportBlocksToChunk hdl firstHeight chunkSize lastFinalizationRecordIndex getBlockAt = ebtc firstHeight 0 lastFinalizationRecordIndex
   where
-    ebtc height count =
-        resizeOnResized (readFinalizedBlockAtHeight height) >>= \case
-            Nothing -> return count
-            Just sb | NormalBlock normalBlock <- sbBlock sb -> do
-                let serializedBlock =
-                        runPut $ putVersionedBlock (protocolVersion @pv) normalBlock
-                    len = fromIntegral $ BS.length serializedBlock
+    ebtc height count lastFinRecIdx =
+        getBlockAt height >>= \case
+            Nothing -> return (count, lastFinRecIdx)
+            Just (serializedBlock, finalizationRecordIndexM) -> do
+                let len = fromIntegral $ BS.length serializedBlock
                 liftIO $ do
                     BS.hPut hdl $ runPut $ putWord64be len
                     BS.hPut hdl serializedBlock
-                forM_ (blockFields (sbBlock sb)) $ \bf ->
-                    case blockFinalizationData bf of
-                        BlockFinalizationData fr ->
-                            dbsLastFinIndex .= finalizationIndex fr
-                        NoFinalizationData -> return ()
+                let lastFinRecIdx' = fromMaybe lastFinRecIdx finalizationRecordIndexM
```

Comment:
I'm not sure I understand this. So you are looping here over finalized blocks in the database.

So if at the end you'd get a Nothing for finalization index (why does that happen?) then you'd return `0` from this function. 

I guess that in V0 consensus you'd always have a finalization index, so this will be fine.

And in consensus V1 you'd never have it, and it also does not matter since it's not used.

I understand there is a lack of time to do it more cleanly, but could we explain this a bit more at least since it's not so clear from the code itself.

Reply:
We could probably relatively easily change the API a bit so that 
```
GetBlockAt
```
would take an additional type parameter, the finalizationindex.

But I think that'll take you a bit too much time, so this is just a thought for now.

Reply:
I have added some documentation.

Reply:
.. And refactored a bit.